packages/integration-objects/srv/integration-objects/simple-exchange-rate-type-service.js

```diff
@@ -0,0 +1,210 @@
+const { fetchCurrentAttributeValues } = require('../utils/FetchCurrentStateForAuditLog');
+
+module.exports = srv => {
+  srv.before('READ', 'ExchangeRateTypes', beforeRead);
+  srv.before('CREATE', 'ExchangeRateTypes', beforeCreate);
+  srv.before('UPDATE', 'ExchangeRateTypes', beforeUpdate);
+  srv.before('DELETE', 'ExchangeRateTypes', beforeDelete);
+
+  srv.before('CREATE', 'ExchangeRateTypeDescriptions', beforeCreateForRateTypeDescriptions);
+  srv.before('UPDATE', 'ExchangeRateTypeDescriptions', beforeUpdateForRateTypeDescriptions);
+  srv.before('DELETE', 'ExchangeRateTypeDescriptions', beforeActiveDeleteForExchangeRateTypeDescription);
+
+  srv.on('CREATE', 'ExchangeRateTypes', writeAuditLogForRateTypes);
+  srv.on('DELETE', 'ExchangeRateTypes', writeAuditLogForRateTypes);
+  srv.on('UPDATE', 'ExchangeRateTypes', writeAuditLogForRateTypes);
+
+  const { ExchangeRateTypes, ExchangeRateTypeDescriptions } = srv.entities;
+
+  function beforeRead(req) {
+    try {
+      checkAndAppendTenantIdFilterForReadEvent(req);
+    } catch (err) {
+      throw new ValidationError(err.message, err.code);
+    }
+  }
+
+  async function beforeCreate(req) {
+    try {
+      await validatePayload(req);
+    } catch (err) {
+      throw new ValidationError(err.message, err.code);
+    }
+  }
+
+  async function beforeUpdate(req) {
+    try {
+      await validatePayload(req);
+    } catch (err) {
+      throw new ValidationError(err.message, err.code);
+    }
+  }
+
+  async function validatePayload(req) {
+    validateTenantIdInPayload(req);
+    validateFieldsForExchangeRateTypes(req.data);
+    await checkUniquenessForPrimaryKeys(req);
+  }
+
+  function beforeDelete(req) {
+    try {
+      validateDelete(req);
+    } catch (err) {
+      throw new ValidationError(err.message, err.code);
+    }
+  }
+
+  async function beforeCreateForRateTypeDescriptions(req) {
+    try {
+      await validateDescPayload(req);
+    } catch (err) {
+      throw new ValidationError(err.message, err.code);
+    }
+  }
+
+  async function beforeUpdateForRateTypeDescriptions(req) {
+    try {
+      await validateDescPayload(req);
+    } catch (err) {
+      throw new ValidationError(err.message, err.code);
+    }
+  }
+
+  async function validateDescPayload(req) {
+    validateFieldsForExchangeRateTypeDescription(req.data, req.params);
+    await checkForValidityOfKey(req, ExchangeRateTypes);
+    await checkUniquenessForPrimaryKeysForExchangeRateTypeDescForUpdate(req);
+  }
+
+  async function checkUniquenessForPrimaryKeys(req) {
+    const record = req.data;
+    const affectedRows = await fetchExistingRateTypes(ExchangeRateTypes, record, req);
+    checkForDuplicateRateTypeObjects(affectedRows, req);
+  }
+
+  async function checkUniquenessForPrimaryKeysForExchangeRateTypeDescForUpdate(req) {
+    const record = req.data;
+    const affectedRows = await fetchExistingRateTypeDesc(ExchangeRateTypeDescriptions, record, req);
```

Comment:
The third parameter is unnecessary here. The following would be more concise:
```suggestion
    const affectedRows = await fetchExistingRateTypeDesc(ExchangeRateTypeDescriptions, req);
```

Then in `fetchExistingRateTypeDesc` you could call `req.record`. The same applies to `fetchExistingTexts` and `fetchExistingRateTypes`.